File: backend/src/fablespace_api/core/relationship_graph.py
"""Owner-governed tavern/character relationship graph domain helpers."""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import Any, Iterable

NODE_TYPES = {"tavern", "character"}
BEHAVIOR_TYPES = {"friendly", "allied", "neutral", "rival", "hostile"}
STRENGTH_PRESETS = {"weak", "normal", "strong"}
GOVERNANCE_MODES = {"manual", "assisted", "delegated_ai", "system_ai"}
EDGE_STATUSES = {"pending", "confirmed", "rejected", "disabled"}

_STRENGTH_MULTIPLIERS = {
    "weak": 0.25,
    "normal": 0.5,
    "strong": 0.8,
}

_SPECIFICITY_RANKS = {
    ("tavern", "tavern"): 1,
    ("tavern", "character"): 2,
    ("character", "tavern"): 2,
    ("character", "character"): 3,
}


def _require_choice(value: str, allowed: set[str], field_name: str) -> str:
    if value not in allowed:
        allowed_text = ", ".join(sorted(allowed))
        raise ValueError(f"{field_name} must be one of: {allowed_text}")
    return value


def _clamp_affinity(value: float) -> float:
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        numeric = 0.0
    return max(0.0, min(1.0, numeric))


def _normalize_non_negative(value: float) -> float:
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        numeric = 0.0
    return max(0.0, numeric)


@dataclass(slots=True)
class RelationshipEdge:
    """Canonical owner/source-side relationship stance between two graph nodes."""

    id: str
    source_node_type: str
    source_node_id: str
    target_node_type: str
    target_node_id: str
    behavior_type: str
    source_owner_id: str = ""
    source_space_id: str = ""
    target_owner_id: str = ""
    target_space_id: str = ""
    display_name: str = ""
    description: str = ""
    strength_preset: str = "normal"
    status: str = "pending"
    governance_mode: str = "manual"
    confirmed_by: str = ""
    confirmed_by_type: str = ""
    created_at: str = ""
    updated_at: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.source_node_type = _require_choice(self.source_node_type, NODE_TYPES, "source_node_type")
        self.target_node_type = _require_choice(self.target_node_type, NODE_TYPES, "target_node_type")
        self.behavior_type = _require_choice(self.behavior_type, BEHAVIOR_TYPES, "behavior_type")
        self.strength_preset = _require_choice(self.strength_preset, STRENGTH_PRESETS, "strength_preset")
        self.status = _require_choice(self.status, EDGE_STATUSES, "status")
        self.governance_mode = _require_choice(self.governance_mode, GOVERNANCE_MODES, "governance_mode")
        self.metadata = dict(self.metadata or {})

    @property
    def is_confirmed_enabled(self) -> bool:
        return self.status == "confirmed"


@dataclass(slots=True)
class RelationshipProjection:
    """Visitor-private dual-axis relation to one tavern or character node."""

    visitor_id: str
    node_type: str
    node_id: str
    space_id: str = ""
    affinity: float = 0.0
    hostility: float = 0.0
    last_event_at: str = ""
    updated_at: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.node_type = _require_choice(self.node_type, NODE_TYPES, "node_type")
        self.affinity = _clamp_affinity(self.affinity)
        self.hostility = _normalize_non_negative(self.hostility)
        self.metadata = dict(self.metadata or {})


def strength_multiplier(value: str) -> float:
    """Return the deterministic multiplier for a closed strength preset."""

    return _STRENGTH_MULTIPLIERS[_require_choice(value, STRENGTH_PRESETS, "strength_preset")]


def apply_positive_effect(
    projection: RelationshipProjection,
    amount: float,
    *,
    cap: float = 1.0,
) -> RelationshipProjection:
    """Increase affinity up to a friendly cap without reducing hostility."""

    effect = _normalize_non_negative(amount)
    cap_value = _clamp_affinity(cap)
    current = _clamp_affinity(projection.affinity)
    if current >= cap_value or effect == 0:
        new_affinity = current
    else:
        new_affinity = min(cap_value, current + effect)
    return replace(
        projection,
        affinity=round(new_affinity, 10),
        hostility=round(_normalize_non_negative(projection.hostility), 10),
    )


def apply_negative_effect(projection: RelationshipProjection, amount: float) -> RelationshipProjection:
    """Drain affinity first, then convert any remainder into hostility."""

    effect = _normalize_non_negative(amount)
    current_affinity = _clamp_affinity(projection.affinity)
    current_hostility = _normalize_non_negative(projection.hostility)
    if effect == 0:
        return replace(projection, affinity=current_affinity, hostility=current_hostility)

    drained = min(current_affinity, effect)
    remainder = effect - drained
    return replace(
        projection,
        affinity=round(max(0.0, current_affinity - drained), 10),
        hostility=round(current_hostility + remainder, 10),
    )


def specificity_rank(edge: RelationshipEdge) -> int:
    """Rank edge specificity; higher means the edge wins propagation resolution."""

    return _SPECIFICITY_RANKS.get((edge.source_node_type, edge.target_node_type), 0)


def select_effective_edge(edges: Iterable[RelationshipEdge]) -> RelationshipEdge:
    """Pick the edge whose specificity should explain one propagation effect."""

    candidates = list(edges)
    if not candidates:
        raise ValueError("at least one relationship edge is required")
    return sorted(
        candidates,
        key=lambda edge: (
            specificity_rank(edge),
            strength_multiplier(edge.strength_preset),
            edge.id,
        ),
        reverse=True,
    )[0]


@dataclass(slots=True)
class RelationshipPropagationEvent:
    """One direct visitor relationship change used as propagation input."""

    visitor_id: str
    source_node_type: str
    source_node_id: str
    source_space_id: str = ""
    source_owner_id: str = ""
    delta_axis: str = "affinity"
    delta_amount: float = 0.0
    character_influence_weight: float = 0.0
    source_event_id: str = ""
    source_message_ids: tuple[str, ...] = ()
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.source_node_type = _require_choice(self.source_node_type, NODE_TYPES, "source_node_type")
        if self.delta_axis not in {"affinity", "hostility"}:
            raise ValueError("delta_axis must be one of: affinity, hostility")
        self.delta_amount = _normalize_non_negative(self.delta_amount)
        self.character_influence_weight = _normalize_non_negative(self.character_influence_weight)
        self.source_message_ids = tuple(self.source_message_ids or ())
        self.metadata = dict(self.metadata or {})


@dataclass(slots=True)
class RelationshipPropagationResult:
    """Summary of one projection updated by propagation."""

    edge_id: str
    target_node_type: str
    target_node_id: str
    affinity: float
    hostility: float
    reason: str = "edge"

File: apps/web/app/lib/relationship-graph.js
export const RELATIONSHIP_NODE_TYPES = [
  { id: "space", label: "空间" },
  { id: "character", label: "角色" },
]

export const RELATIONSHIP_BEHAVIOR_TYPES = [
  { id: "friendly", label: "友好" },
  { id: "allied", label: "同盟" },
  { id: "neutral", label: "中立" },
  { id: "rival", label: "竞争" },
  { id: "hostile", label: "敌对" },
]

export const RELATIONSHIP_STRENGTH_PRESETS = [
  { id: "weak", label: "弱" },
  { id: "normal", label: "标准" },
  { id: "strong", label: "强" },
]

export const RELATIONSHIP_GOVERNANCE_MODES = [
  { id: "manual", label: "店主手动" },
  { id: "assisted", label: "辅助建议" },
  { id: "delegated_ai", label: "委托 AI" },
  { id: "system_ai", label: "系统 AI" },
]

export const RELATIONSHIP_EDGE_STATUSES = [
  { id: "pending", label: "待确认" },
  { id: "confirmed", label: "已确认" },
  { id: "rejected", label: "已拒绝" },
  { id: "disabled", label: "已停用" },
]

const NODE_TYPE_IDS = new Set(RELATIONSHIP_NODE_TYPES.map((item) => item.id))
const BEHAVIOR_TYPE_IDS = new Set(RELATIONSHIP_BEHAVIOR_TYPES.map((item) => item.id))
const STRENGTH_PRESET_IDS = new Set(RELATIONSHIP_STRENGTH_PRESETS.map((item) => item.id))
const GOVERNANCE_MODE_IDS = new Set(RELATIONSHIP_GOVERNANCE_MODES.map((item) => item.id))
const EDGE_STATUS_IDS = new Set(RELATIONSHIP_EDGE_STATUSES.map((item) => item.id))

function text(value) {
  return typeof value === "string" ? value.trim() : String(value ?? "").trim()
}

function metadataObject(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? { ...value } : {}
}

function choose(value, allowed, fallback) {
  const normalized = text(value).toLowerCase().replaceAll("-", "_")
  return allowed.has(normalized) ? normalized : fallback
}

export function normalizeRelationshipNodeType(value, fallback = "space") {
  return choose(value, NODE_TYPE_IDS, fallback)
}

export function normalizeRelationshipBehaviorType(value, fallback = "friendly") {
  return choose(value, BEHAVIOR_TYPE_IDS, fallback)
}

export function normalizeRelationshipStrengthPreset(value, fallback = "normal") {
  return choose(value, STRENGTH_PRESET_IDS, fallback)
}

export function normalizeRelationshipGovernanceMode(value, fallback = "manual") {
  return choose(value, GOVERNANCE_MODE_IDS, fallback)
}

export function normalizeRelationshipEdgeStatus(value, fallback = "confirmed") {
  return choose(value, EDGE_STATUS_IDS, fallback)
}

export function normalizeRelationshipEdge(value = {}, defaults = {}) {
  const sourceSpaceId = text(value.source_space_id || defaults.source_space_id)
  const targetSpaceId = text(value.target_space_id || defaults.target_space_id)
  const sourceNodeType = normalizeRelationshipNodeType(value.source_node_type, defaults.source_node_type || "space")
  const targetNodeType = normalizeRelationshipNodeType(value.target_node_type, defaults.target_node_type || "space")
  return {
    id: text(value.id || defaults.id),
    source_owner_id: text(value.source_owner_id || defaults.source_owner_id),
    source_space_id: sourceSpaceId,
    source_node_type: sourceNodeType,
    source_node_id: text(value.source_node_id || defaults.source_node_id || (sourceNodeType === "space" ? sourceSpaceId : "")),
    target_owner_id: text(value.target_owner_id || defaults.target_owner_id),
    target_space_id: targetSpaceId,
    target_node_type: targetNodeType,
    target_node_id: text(value.target_node_id || defaults.target_node_id || (targetNodeType === "space" ? targetSpaceId : "")),
    behavior_type: normalizeRelationshipBehaviorType(value.behavior_type, defaults.behavior_type || "friendly"),
    display_name: text(value.display_name || defaults.display_name),
    description: text(value.description || defaults.description),
    strength_preset: normalizeRelationshipStrengthPreset(value.strength_preset, defaults.strength_preset || "normal"),
    status: normalizeRelationshipEdgeStatus(value.status, defaults.status || "confirmed"),
    governance_mode: normalizeRelationshipGovernanceMode(value.governance_mode, defaults.governance_mode || "manual"),
    confirmed_by: text(value.confirmed_by || defaults.confirmed_by),
    confirmed_by_type: text(value.confirmed_by_type || defaults.confirmed_by_type || "owner"),
    perspective_scope: text(value.perspective_scope || defaults.perspective_scope || "source_owner"),
    created_at: text(value.created_at || defaults.created_at),
    updated_at: text(value.updated_at || defaults.updated_at),
    metadata: metadataObject(value.metadata || defaults.metadata),
  }
}

export function normalizeRelationshipEdgeDraft(value = {}, options = {}) {
  const draft = normalizeRelationshipEdge(value, options)
  const fallbackSourceCharacterId = text(options.fallback_source_character_id)
  if (draft.source_node_type === "space") {
    draft.source_node_id = draft.source_space_id
  } else if (!draft.source_node_id) {
    draft.source_node_id = fallbackSourceCharacterId
  }
  if (draft.target_node_type === "space") {
    draft.target_node_id = draft.target_space_id
  }
  return draft
}

export function createRelationshipEdgeDraftSeed(space = {}, characters = []) {
  const spaceId = text(space.id)
  const firstCharacterId = Array.isArray(characters) ? text(characters[0]?.id) : ""
  return normalizeRelationshipEdgeDraft(
    {
      source_space_id: spaceId,
      source_node_type: firstCharacterId ? "character" : "space",
      source_node_id: firstCharacterId || spaceId,
      target_space_id: "",
      target_node_type: "space",
      target_node_id: "",
      behavior_type: "friendly",
      display_name: "",
      description: "",
      strength_preset: "normal",
      status: "confirmed",
      governance_mode: "manual",
      confirmed_by_type: "owner",
    },
    {
      source_space_id: spaceId,
      fallback_source_character_id: firstCharacterId,
    },
  )
}

export function normalizeRelationshipEdges(value, defaults = {}) {
  return Array.isArray(value) ? value.map((item) => normalizeRelationshipEdge(item, defaults)) : []
}

export function relationshipOptionLabel(options, value, fallback = "") {
  return options.find((item) => item.id === value)?.label || fallback || value || ""
}

export function isCrossOwnerRelationshipPerspective(edge, ownerId = "") {
  const currentOwnerId = text(ownerId)
  if (!currentOwnerId) return false
  return text(edge?.target_owner_id) !== "" && text(edge?.target_owner_id) !== currentOwnerId
}
